### PR TITLE
Shoring up the exclusivity diagnostics.

### PR DESCRIPTION
**Pull request by atrick**
Make sure that various SIL patterns can't create holes in the exclusivity model.

### CODE REVIEW

**Comment by atrick** [comment]
@devincoughlin can you take a quick look. Rather than parsing the diff, I think it's best just to read the code that handles noescape closures. It's all in one section of the file, about 250 lines, from 
```
+// =============================================================================
 +// The data flow algorithm that drives diagnostics.
```
to
```
+// =============================================================================
 +// Verification
```

**Comment by atrick** [comment]
@swift-ci test.

**Comment by atrick** [comment]
@swift-ci test source compatibility.

File: lib/SILOptimizer/Mandatory/DiagnoseStaticExclusivity.cpp [hunk 1]
```diff
@@ -718,20 +734,20 @@ static void checkForViolationWithCall(
...
     // Are there any accesses in progress at the time of the call?
-    if (AccessIt == Accesses.end())
+    if (AccessIt == State.Accesses->end())
       continue;
 
     const AccessInfo &Info = AccessIt->getSecond();
     if (auto Conflict = findConflictingArgumentAccess(AS, Storage, Info)) {
-      Conflict->setAlwaysDiagnoseAsWarning(DiagnoseAsWarning);
-      ConflictingAccesses.push_back(*Conflict);
+      Conflict->setAlwaysDiagnoseAsWarning(/*DiagnoseAsWarning=*/false);
```
Comment (shajrawi) [inline lib/SILOptimizer/Mandatory/DiagnoseStaticExclusivity.cpp]: I agree with you that it should never be a warning, but, why keep the callee function as one taking a bool? Looking at the code base this is the only call site - can simplify the callee as well.

**Review by shajrawi — CHANGES_REQUESTED** [review]
Overall LGTM, but, please update consider updating the `setAlwaysDiagnoseAsWarning` function

File: lib/SILOptimizer/Mandatory/DiagnoseStaticExclusivity.cpp (hunk 1 above)
Comment (atrick) [reply lib/SILOptimizer/Mandatory/DiagnoseStaticExclusivity.cpp]: Fixed.

**Comment by atrick** [comment]
@swift-ci test.

**Review by shajrawi — APPROVED** [review]
LGTM!

**Pull request merged**